### ISSUE TITLE
FR: Add 'Invert Images' option to .CBZ and .PDF file formats 

### ISSUE THREAD

**Issue by Neko-Kunshu** [koreader/koreader#9899]
**Does your feature request involve difficulty completing a task? Please describe.**
No 'Invert Images' option for .CBZ and .PDF file formats necessitates constant manual toggling of 'Night mode' feature whenever switching between reading, for example, EPUBs (where 'Invert Images' option is available) to reading comic book/manga file formats. 

**Describe the solution you'd like**
Add 'Invert Images' option for .PDF and .CBZ file formats.

**Additional context**
Here's how reading a .PDF or .CBZ comic book/manga looks like with 'Night mode' turned on as of now:
![Screenshot_20221211-130713](https://user-images.githubusercontent.com/85441970/206900048-637aaf9a-cb49-461e-ac06-42a17aa2dc2a.jpg)
![Screenshot_20221211-130716](https://user-images.githubusercontent.com/85441970/206900056-09a3617e-9189-4f1e-bfc3-cc7b3aef7e50.jpg)

And here's how it will look like with requested feature implemented:
![Screenshot_20221211-130725](https://user-images.githubusercontent.com/85441970/206900096-66196051-3cdc-4c80-a08f-fe3b35824d04.jpg)
![Screenshot_20221211-130728](https://user-images.githubusercontent.com/85441970/206900102-c6959fc9-98dd-42f1-9d86-d691b972d1c0.jpg)

**Comment by Neko-Kunshu** [comment]
So... Is this a stupid suggestion or... a reasonable one?

**Comment by poire-z** [comment]
Be patient for answers :)
(I don't know much about PDF, but I'd say it's probably harder to distinguish images from text in PDF, and to copilot the PDF rendering engine - which is somehow easier in EPUB, but still needs us to copilot the EPUB rendering engine from the frontend side when nightmode images.)

**Comment by Neko-Kunshu** [comment]
Stop reading the comic and comment something, please! 🥺

**Comment by Frenzie** [comment]
PDFs are built up out of many layers and elements. For most or at least many PDF-based documents a suggestion like this couldn't really work out, including comics in PDF. 

For CBZ it's at least theoretically possible because every page is simply an image.

If the sample comic is a PDF it would probably work for that specific one because it's likely to be very CBZ-like in its construction, presumably being little but a series of scanned images.

However, instead of "invert images" it should be more reasonably possible to have "invert document" without worrying too much about the image part. For the moment I'd suggest to shelve that idea because a big upgrade of MuPDF/k2pdfopt is hopefully incoming in the nearby future.

**Comment by NiLuJe** [comment]
Also, it would be nearly identical to just disabling nightmode?

**Comment by Neko-Kunshu** [comment]
> Also, it would be nearly identical to just disabling nightmode?

You're missing my point. Disabling Night mode affects ALL files in KOReader, and not only files, but KOReader interface. What I mean to achieve by suggesting this feature is ability to sort of disable the image color inversion (which happens when Night mode is enabled) on per-file basis for PDF and CBZ, so that each time I'm reading a, say, comic, I won't have the need to manually toggle Night mode on or off.

**Comment by NiLuJe** [comment]
I meant on a per-document basis, and possibly on a technical level that it might be easier to just disable nightmode than invert the buffer (I don't recall how we blit CBZs).

Granted, touching nightmode would affect the UI, too, so the invert approach would be preferred ;).

**Comment by NiLuJe** [comment]
In any case, unlikely to see any traction as far as I'm concerned because PDF + nightmode is a double whammy of meh ;).

But, depending on how we blit CBZs, it might be trivial-ish to implement (except the bottom menu is kinda crowded, IIRC).

**Comment by criztovyl** [comment]
I'm interested in this, running into this with CBZ's extracted from EPUBs. I looked around in source code, it seems CBZ uses muPDF (whike EPUB uses crengine).

Maybe for CBZs the full page (which should be only a single image, correct?) could be inverted (using muPDF `fz_invert_matrix`?) in the following place?

https://github.com/koreader/koreader-base/blob/master/ffi/mupdf.lua#L625-674

**Comment by NiLuJe** [comment]
No, we definitely don't want to handle the page separately from the UI.

We already have the necessary blitting tooling, I just have subzero interest in this myself ;).

**Comment by criztovyl** [comment]
> don't want to handle the page separately from the UI.

What do you mean by that?

My idea was to have it similar to crengine, where the `invert_images` flag is passed down in `CreDocument:drawCurrentView`'s, via libkoreader-cre / cre.cpp, where the flag is set on`LvDrawBuf`.
For muPDF i would pass down the inversion flag to `page_mt.__index:draw_new`, where depending on the flag `fz_invert_matrix` is used to invert the page (which in case of CBZ is a single image).

**Comment by poire-z** [comment]
I think he meant:
- in crengine, we may have images AND text, and crengine knows (we were familiar enough with crengine to make it know :) what is text and what is image, and can invert only some of them (images), so your text is still non-inverted, as expected
- in MuPDF, we most probably don't know (or don't want to get too familiar with it to be able to know :) what is image, and what is text, and some internal images may be text... and in the end, we only get a single image of the full screen. And we could invert that single full screen image with the UI code (we have code for that, "_the necessary blitting tooling_"), instead of hacking MuPDF to have it do it for us.

**Comment by Frenzie** [comment]
We do know a lot about images in MuPDF in theory, but given the high quantity of false positives I certainly wouldn't bother.

But for CBZ as opposed to PDF I don't understand what the point would even be? It's just a 1 on 1 thing.

**Comment by NiLuJe** [comment]
Basically what @poire-z said ;).

Added to the fact that, even if there wasn't an image vs. bg vs. text issue, there'd still be a "page buffer" vs. UI issue where one would be inverted and not the other. (Unless I'm missing the point, which is quite possible, because I hate nihgtmode with a fiery passion).

**Comment by criztovyl** [comment]
I still struggle to grasp the page-buffer vs UI issue.

My understanding of night mode is that after the underlying library (crengine, mupdf) has rendered the page into a buffer it is blitted to the UI and the result is inverted, correct?

for EPUB comics this implementation of night mode is not a problem, crdocuments can invert images, so if everything is inverted, images are inverted again and thus are not inverted anymore.

For CBZ comics the implementation of night mode is a problem, pdfdocuments cannot be asked to invert images, so they get inverted only once.

The idea now is to add a nightmode flag to pdfdocuments that only is set for CPZs.

This would mean the whole page buffer for CPZs will be inverted again, and images would be correct again. If there actually is a page that is text and not image, well, it will still be readable, just also double inverted and as such should be equivalnet to having to night mode activated.

this would mean the UI elements would still be inverted once, appearing dark, only the page will be inverted twice and as such stay light.

And, out of curiosity, where is the hatred of nightmode rooted in?

**Comment by NiLuJe** [comment]
> where is the hatred of nightmode rooted in

In the fact that it looks like absolute ass unless you're using a latest gen device with proper eclipse waveform modes (bonus points if the frontlight is smarter, too).

**Comment by NiLuJe** [comment]
In any case, we have "in flight" invert blitters, so you can do that for nearly free via those instead of adding an `fz_invert_pixmap` pass somewhere inside MuPDF.

(And doing it in front means it'll more easily survive backend updates).

**Comment by Nuihc88** [comment]
Any update on this? I was in the process of filling out a separate Feature Request issue on this problem, when i found this one... Here is what i had written up:

**Does your feature request involve difficulty completing a task? Please describe.**
When the Night Mode is enabled, picture-as-page formats such as '.cbz', '.cbr' & '.cbt' become inverted and thus borderline unviewable in the case of Comics & Manga.

**Describe the solution you'd like**
I would like for the images in picture-as-page formats to be displayed normally, while still having all GUI elements such as margins, notifications and menus inverted as usual.

It could be done by default for certain formats only or through a settings toggle like 'copt_nightmode_images' currently found in 'settings.reader.lua' or through a GUI settings menu. Any of those options would work for me.

**Describe alternatives you've considered**
I have looked into various CSS workarounds and outdated unofficial patches to get around the issue, but none have worked well for me with up to date release version of koreader.

As a temporary workaround i have now configured one corner of the touch screen on my Kindle PW3 to toggle Night Mode On/Off and hidden the status and progress bars behind a long press on another corner, but due to requiring a touch screen this solution wouldn't work for my K3 & K4NT.

**Additional context**
I tried toggling the 'copt_nightmode_images' entry in the 'settings.reader.lua' file, but currently that doesn't seem to have any effect on these formats.

**Comment by noxhirsch** [comment]
@Nuihc88 I've finally managed to create a working solution (with a little bit help of AI) - even if it's a very suboptimal workaround. Maybe it will help you anyway - until someone with real programming skills comes along and implements a proper solution ;)
You can add this function to `ffi/mupdf.lua` (I've added it after the end of the function `run_page`):

```
local function run_page_nightmode(page, pixmap, ctm)
    M.fz_clear_pixmap_with_value(page.ctx, pixmap, 0xff)

    local dev = W.mupdf_new_draw_device(page.ctx, nil, pixmap)
    if dev == nil then merror(page.ctx, "cannot create draw device") end

    local ok = W.mupdf_run_page(page.ctx, page.page, dev, ctm, nil)

    if page.number > 1 then
        local w = M.fz_pixmap_width(page.ctx, pixmap)
        local h = M.fz_pixmap_height(page.ctx, pixmap)
        local n = M.fz_pixmap_components(page.ctx, pixmap)
        local samples = M.fz_pixmap_samples(page.ctx, pixmap)
        if n == 4 then
            for i = 0, w * h - 1 do
                local p = i * 4
                samples[p+0] = 255 - samples[p+0]
                samples[p+1] = 255 - samples[p+1]
                samples[p+2] = 255 - samples[p+2]
            end
        end
    end

    M.fz_close_device(page.ctx, dev)
    M.fz_drop_device(page.ctx, dev)
    if ok == nil then merror(page.ctx, "could not run page") end
end
```
Then replace `run_page` in the function `page_mt.__index:draw_new` with `run_page_nightmode`. You have to redo that after every update.

The first page does not get inverted, otherwise the cover would be inverted in the mosaic cover browser.

**Comment by Nuihc88** [comment]
@noxhirsch I tried your workaround on my 'Kindle Paperwhite 3' with 'KOReader 2025.04', but couldn't get it to work.
[Here is a zipped copy of my modified ffi/mupdf.lua](https://github.com/user-attachments/files/21379952/mupdf.lua.zip)

**Comment by noxhirsch** [comment]
@Nuihc88 Looks good to me. But I forgot to mention: Since this is a backend change, it has no effect on already cached pages. Have you tried turning a few pages or opening a new book?

**Comment by Nuihc88** [comment]
@noxhirsch I tried clearing my cache, manually deleting all files in my cache folder, setting view mode from 'continuous' to 'page', disabling some plugins and then opening a newly added .cbz-file. No change.

**Comment by noxhirsch** [comment]
@Nuihc88 You could try to remove the `if page.number > 1 then` and the corresponding `end` to check if that changes anything. I'm running version 2025.04-138 - maybe an update helps if something relevant has changed since version 2025.04. 
If that doesn't help either, I'm afraid I don't have any other ideas. I barely managed to get it working for me. But I don't know enough about the LUA and the inner workings of KOReader to do remote debugging ;) I also don't know what the differences are between the various hardware platforms.

**Comment by microfloraa** [comment]
Bumping this issue, I have a profile set to automatically disable Night Mode when I open a .cbz file to prevent image color inversion (and another profile that automatically re-enables Night Mode when I go back to Home), but it would still be really nice to have Night Mode menus while reading .cbz documents without images getting inverted.

**Comment by Euphoriyy** [comment]
I've made a patch to invert CBZs/PDFs in night mode with an option [here](https://github.com/Euphoriyy/KOReader.patches/blob/master/2-invert-document.lua). if anyone tries it, let me know how well it works for you. I'm also making a PR based on it which hopefully can be merged soon.

Would it be better for this option to be under the contrast settings or general settings (where "Invert Images" is)?

<img width="500" height="auto" alt="Image" src="https://github.com/user-attachments/assets/6201938e-a2cc-4211-ba24-5d5548f09ed4" />

**Comment by noxhirsch** [comment]
Hey @Euphoriyy, thank you very much! I tested the patch and it works great for me! :)

>Would it be better for this option to be under the contrast settings or general settings (where “Invert Images” is)?

Since it's a color setting, I would tend to look for it under the contrast settings - so it fits well there for me.

**Comment by Euphoriyy** [comment]
Thanks for the feedback @noxhirsch, it's nice to know that it worked well! I made the PR for it now as well.
